### PR TITLE
fix(persistentShell): authoritative post-kill stdout capture

### PR DESCRIPTION
**Pull request by kylejryan**
## Summary

Silent stdout loss in `PersistentShell` on macOS (and Alpine / minimal containers) is corrupting agent behavior across the APEX TUI, the `pentest` CLI, and every Evalgate-driven frontier eval that runs APEX on a Mac. Short commands (`echo`, `pwd`, `ls -la`, most recon probes) return empty stdout, which [`executeCommand.ts`](../blob/canary/src/core/agents/offSecAgent/tools/executeCommand.ts#L69) turns into `"(no output)"`. The agent believes its recon returned nothing, abandons real attack paths, and hallucinates findings from whatever did come through (browser tool, HTTP tool, stderr).

This PR makes final-stdout capture authoritative and platform-independent by adding a post-kill `cat` phase that bypasses `tail`/libc buffering entirely.

## The bug

The current wrapper streams per-command stdout to Node by running a background `tail -f` of a tempfile:

```bash
stdbuf -oL tail -n +1 -f -s 0.05 "$__APEX_OUT" 2>/dev/null &
__APEX_TAIL=\$!
{ <command> } </dev/null >\"\$__APEX_OUT\" 2>\"\$__APEX_ERR\"
__APEX_EC=\$?
sleep 0.1
kill \"\$__APEX_TAIL\" 2>/dev/null
wait \"\$__APEX_TAIL\" 2>/dev/null
```

Correctness depends on two assumptions:

1. `stdbuf -oL` actually line-buffers `tail`'s stdout pipe.
2. `tail` flushes everything before we kill it.

On macOS under SIP, **both fail**. SIP strips \`DYLD_INSERT_LIBRARIES\` from protected system binaries, including \`/usr/bin/tail\`, so \`stdbuf\`'s \`LD_PRELOAD\`/\`libstdbuf\` trick is a silent no-op. Tail block-buffers its stdout pipe (~8 KiB default), the \`kill\` fires ~100 ms later, tail's buffer is discarded with the process, and the parent Node handler sees only the exit marker. \`cmd.stdout\` ends up \`\"\"\`.

Stderr survives because it's \`cat\`'d synchronously after the command, bypassing tail.

### Platform scope

| Platform | Effective \`stdbuf\`? | \`tail\` stdout | Bug? | Why |
|---|---|---|---|---|
| macOS (Intel / Apple Silicon), SIP on | No | Block-buffered | **Yes, when \`stdbuf\` is on PATH** (brew coreutils) | SIP strips \`DYLD_INSERT_LIBRARIES\` from \`/usr/bin/tail\` |
| macOS with no \`stdbuf\` on PATH | N/A | N/A | No (today) | [#636](https://github.com/pensarai/apex/pull/636) added a \`cat\`-only fallback for this case |
| Linux glibc + coreutils (Debian/Ubuntu/Fedora/Arch, CI, Daytona) | Yes | Line-buffered | No | \`LD_PRELOAD\` works, libstdbuf forces line buffering |
| Linux (musl / Alpine) | Partial | Block-buffered | Often yes | BusyBox \`tail\` ignores libstdbuf; \`libstdbuf\` may not load on musl |
| Linux glibc without coreutils (minimal containers) | No | N/A | No (today, via #636) | \`cat\`-only fallback |
| Inside APEX Daytona sandbox | Yes | Line-buffered | No | glibc + coreutils |
| Windows | N/A | N/A | No | Separate \`cmd.exe\` code path |

Where the bug actually fires today: APEX TUI on a Mac, APEX CLI pentest on a Mac, Evalgate \`runApexFromSource\` spawning APEX on a Mac (the NOCTURNE path), APEX running inside Alpine/BusyBox targets. Never fires inside the Daytona sandbox, which is why benchmark runs don't trip it. Never caught by CI unit tests, which run on Ubuntu/glibc.

### Why [#636](https://github.com/pensarai/apex/pull/636) didn't close the bug

PR #636 made \`hasStdbuf()\` fall back to \`cat \"\$__APEX_OUT\"\` when \`stdbuf\` isn't installed at all. It fixed the \"vanilla macOS dev machine\" case. But users with brew coreutils installed, or any container with partial GNU tooling, still take the \`tail -f\` path — and on those setups \`stdbuf\` is present but **ineffective**, so the original failure still fires. \`hasStdbuf()\` doesn't distinguish \"\`stdbuf\` exists\" from \"\`stdbuf\` is effective against the dynamic linker on this platform\".

## The fix

**Keep \`tail -f\` for live UX; add an authoritative post-kill \`cat\` phase that Node treats as the source of truth.**

After killing the background tail, the wrapper emits:

1. A per-command random cutover marker (\`__APEX_<nonce>__CUTOVER\`) via \`printf\` on bash's real stdout pipe.
2. \`cat \"\$__APEX_OUT\"\` — also through bash's real stdout pipe, with no tail or libc buffering in between.

Node-side, \`onStdoutData\` splits the incoming stream at the cutover marker:

- **Pre-cutover** bytes → fed to \`onData\` for live UX streaming, buffered in \`streamedStdout\` as a fallback for kill-before-cutover paths (timeout, abort, cancel).
- **Post-cutover** bytes → accumulated into \`authoritativeStdout\`. Exit-marker detection runs against this buffer, so the user command's own stdout cannot accidentally trigger early resolution from the pre-cutover phase.

\`PendingCommand\` now has \`streamedStdout\`, \`authoritativeStdout\`, \`cutoverSeen\`, \`cutoverMarker\` in place of a single \`stdout\`. All four fallback-resolve sites (close handler, timeout fallback, abort fallback, \`cancelCurrentCommand\`) return \`authoritativeStdout || streamedStdout || \"(no output)\"\` — so a command killed before the cutover phase still returns whatever tail managed to flush, never worse than today.

The two old wrapper branches (\`hasStdbuf\` → tail-with-stdbuf vs \`!hasStdbuf\` → cat-only) collapse into one unified wrapper. \`stdbuf\` is still used as a hint for better streaming UX when it's effective, but final-stdout correctness no longer depends on it.

### Chunk-boundary safety

An early iteration of the handler had a bug where a cutover marker straddling two Node Buffer reads would be missed (\`chunk.indexOf(marker)\` returns -1 on each chunk alone). The committed handler searches the accumulated \`streamedStdout\` (not just the latest chunk), so markers that arrive split are matched correctly; \`onData\` only receives the new-chunk slice of the pre-cutover bytes, so streamed bytes are never emitted twice.

### Behavior preserved

- \`cd\`, \`export\`, shell functions, aliases still persist (brace-group wrapping unchanged).
- Stdin still redirected from \`/dev/null\` (stdin-hijack fix from #625 unchanged).
- Process-group cleanup on timeout/abort unchanged.
- Exit-code sentinels (124 on timeout, 130 on abort/cancel) unchanged.
- \`MAX_BUFFER\` truncation + \`(stdout truncated)...\` prefix unchanged, now applied to the authoritative buffer.
- Live-streaming UX on Linux glibc unchanged.
- Windows \`cmd.exe\` path completely untouched.

## Risks

| Risk | Likelihood | Mitigation |
|---|---|---|
| Exit-marker false positive inside user output | Negligible | 8 random hex bytes per command; same collision model as today |
| Cutover-marker collision with user output | Negligible | Same nonce; per-command |
| Doubled wire traffic (tail stream + authoritative \`cat\`) | Low | KB-scale for typical commands; Node drains the pipe continuously so nothing accumulates |
| \`cat\` fails mid-phase (tempfile vanishes) | Very low | Fallback returns \`streamedStdout\`; not worse than today |
| Kill before cutover (timeout/abort) | Handled | Forced exit codes unchanged; fallback resolvers return \`streamedStdout\` |

Risks of **not** fixing:

- Every macOS APEX run silently loses stdout for small commands. No user-visible error; \`success: true\` masks it.
- NOCTURNE and every Evalgate frontier eval produce signal dominated by shell brokenness, not agent capability. Current eval scores on Mac runners are contaminated.
- Any deployment on Alpine or minimal containers hits the same bug.
- CI unit tests on Ubuntu/glibc don't catch the failure — silent regression risk forever.

### Added regression tests

[\`src/core/agents/offSecAgent/tools/persistentShell.test.ts\`](../blob/fix/persistent-shell-authoritative-stdout/src/core/agents/offSecAgent/tools/persistentShell.test.ts):

- **\"captures trivial stdout even when tail block-buffers\"** — runs \`echo hello\`, \`pwd\`, \`seq 1 5\` back-to-back. Fails on macOS with \`stdbuf\` on PATH against the pre-patch wrapper (every \`.stdout\` is \`\"(no output)\"\`); passes post-patch.
- **\"captures output across the 8 KiB block-buffer boundary\"** — uses \`head -c 9000 /dev/zero | tr '\\\\0' 'x'\` to emit exactly 9000 \`x\` bytes (well above tail's ~8 KiB default buffer); asserts the full payload survives.
- Existing **\"streams stdout to onData as output arrives\"** — gated on \`!HAS_STDBUF || process.platform === \"darwin\"\`. Comment documents the SIP/DYLD reason and notes that live streaming is a known best-effort UX on darwin; the patch only guarantees final-stdout correctness there. All other platforms continue to enforce the live-streaming property.

### Local verification

```
bun run lint        # clean
bun run format:check  # clean
bun run test src/core/agents/offSecAgent/tools/persistentShell.test.ts
# 10 passed, 1 skipped (streaming test on darwin, by design)
bun run test      # 756 passed, 2 failed
```

The 2 failures in \`src/core/ai/ai.test.ts\` are pre-existing (\`AI_APICallError\` against real Anthropic) — confirmed to reproduce on \`canary\` with the patch stashed, unrelated to this change. Similarly, the single \`src/tui/utils/paste.ts\` \`tsc\` error is pre-existing on \`canary\`.

### Suggested post-merge validation

- Run APEX TUI operator mode on a Mac (\`bun run dev\`) against a local target. Confirm recon commands no longer return \`\"(no output)\"\` by grepping the session trace:
  \`\`\`
  rg -c '\"\\\\(no output\\\\)\"' ~/.pensar/sessions/<id>/subagents/*.trace.jsonl
  \`\`\`
  Pre-patch on Mac: tens to hundreds within the first minute. Post-patch: single digits (only commands that legitimately produce no output).
- Re-run NOCTURNE via Evalgate on a Mac and compare \`(no output)\` counts against baseline. Expect > 90% reduction in \`(no output)\` and a corresponding increase in real shell-based recon reaching the target (e.g. \`/debug/\*\` hits on Meridian rising from 0 to dozens).

### CI matrix

Current CI (\`.github/workflows/ci.yml\`) runs on Linux glibc — the patched correctness tests pass there today. Suggested follow-up (not in this PR): add an Alpine job so \`docker run --rm -v \"\$PWD\":/a -w /a oven/bun:1-alpine bun test\` gets exercised — it would have caught this class of bug pre-emptively.

## Follow-up: fallback-path marker leak (commit 2)

After the initial authoritative-stdout fix landed, a NOCTURNE re-run surfaced a secondary issue on the Node-side fallback paths. The shell-side patch works for the happy path, but the timeout / abort / close / `cancelCurrentCommand` resolvers were returning the raw `streamedStdout` accumulator verbatim — which still contained the per-command `__APEX_<nonce>__CUTOVER` and `__APEX_<nonce>__EXIT_<code>` markers the shell had emitted but Node's chunk handler hadn't yet processed.

### Evidence: NOCTURNE pre-patch vs post-patch vs this follow-up

Comparing `weave_export_evalgate-apex-nocturne-002_2026-04-24.jsonl` (post-initial-patch) to pre-patch baseline `01KPXQXR5D7THJJV9EFRSBHF2E`:

| Metric | Pre-patch (run 001) | Post-patch (run 002) | Remaining |
|---|---|---|---|
| `execute_command` calls | 31 | 74 | — |
| Calls returning real stdout | 0 / 31 (0%) | 50 / 74 (68%) | fixed |
| Calls with clean stdout (no marker leak) | 0 | 47 / 50 (94%) | 3 leaks |
| Calls with marker leaked into stdout | 0 | 3 / 74 | this commit |
| `(no output)` responses | 31 (100%) | 24 (32%) | all genuine timeouts |
|  — of which are stdout-capture bug | 31 | 0 | fixed |
|  — of which are real timeouts | 0 | 24 | out of scope |
| Leaks on commands with `success: true` | 0 | 0 | — |
| Leaks on commands with `error: "Command timed out"` | 0 | 3 / 3 (100%) | this commit |
| Agent progression | stopped ~step 16, zero real recon | reached step 39, ran gobuster 5× + ffuf 10× | fixed |

Three representative leaked records (all from run 002):

| Step | `success` | `error` | Observed stdout (trimmed) |
|---|---|---|---|
| 2 | false | Command timed out | `__APEX_7e680629f7b4a78b__CUTOVER\n/console -> HTTP 404\n/debug -> HTTP 404\n…\n__APEX_7e680629f7b4a78b__EXIT_0\n` |
| 3 | false | Command timed out | `__APEX_da6cc953a56a8138__CUTOVER\n[FOUND] /health -> HTTP 200\n__APEX_da6cc953a56a8138__EXIT_0\n` |
| 7 | false | Command timed out | `__APEX_d4f9e531b2424497__CUTOVER\n__APEX_d4f9e531b2424497__EXIT_143\n` |

Step 2 is the smoking gun: the leaked buffer contains `EXIT_0`, meaning the command actually completed successfully, but Node's timeout timer fired before `onStdoutData` processed the chunk containing the marker — so the fallback resolver ran and returned the raw accumulator instead of the parsed result.

### Root cause

Classic event-loop race. The real trigger is a busy Node loop (AI SDK streaming, telemetry, multiple concurrent tool calls): chunks arrive on the shell's stdout pipe but the `onStdoutData` handler is queued behind other work. Meanwhile, `setTimeout`-driven timeout / abort fallbacks are scheduled at wall-clock deadlines and run when their slot comes up, independent of whether the chunk queue has drained. When the fallback wins the race, `cmd.streamedStdout` still contains the markers verbatim — and the pre-follow-up code returned that buffer untouched.

Hard to force deterministically on an idle test runner (the shell's post-kill wrapper finishes emitting markers within a few hundred ms, almost always beating the 2-second fallback timer in unit tests) — but unambiguous in real NOCTURNE traces, exactly where the event loop is under pressure.

### Fix

Extract a pure helper `extractFallbackStdout(cmd)` that performs the same stripping the happy-path parser does — prefers `authoritativeStdout` when populated, otherwise strips the cutover prefix and exit-marker suffix from `streamedStdout`. Route all five fallback resolve sites through it:

1. `this.proc.on("close", ...)` — shell-death path.
2. Abort path — inner `setTimeout` after SIGKILL escalation.
3. Timeout path — inner `setTimeout` after SIGKILL escalation.
4. `cancelCurrentCommand` SIGKILL-path (Unix).
5. `cancelCurrentCommand` else-branch (Windows).

Node-side only. No shell-script change. No API change. Five call sites, one helper.

### Added regression tests

Real-world race is hard to force deterministically, so the helper is exported for direct unit testing. Seven unit tests in `extractFallbackStdout` describe block exercise the exact leaked-buffer shapes from NOCTURNE traces (steps 2, 3, 7) plus boundary cases — authoritative preferred when populated, both markers absent, cutover-only, exit-only, both adjacent, empty buffer.

Verified as genuine regression guards: with the stripping logic removed, 4 of the 7 new tests fail with the exact shape of leaked buffers observed in production. With the fix, all pass.

Two integration tests (`does not leak nonce markers on timeout`, `does not leak nonce markers on abort`) ride along as a second line of defense — they assert the correct shape of stdout after a real timeout/abort. They pass both with and without the fix in a quiet test environment, because the happy-path parser typically wins the race on an idle runner, but will catch regressions if the helper is bypassed entirely.

### Suggested post-merge validation

After re-running NOCTURNE:

```
RUN=.evalgate/results/nocturne/<new-run>/NOCTURNE-001
rg -c '_CUTOVER|_EXIT_' "$RUN"/pensar-session/subagents/*.trace.jsonl
# Expect 0.
```

### Risks (follow-up only)

| Risk | Likelihood | Mitigation |
|---|---|---|
| Helper strips a marker-lookalike in genuine user output | Negligible | Nonces are 8 random hex bytes per command; zero collision risk. |
| Helper returns `authoritativeStdout` mid-authoritative-phase | Low | That's the right answer — partial authoritative > raw streamed. |
| Fallback fires before tail flushes any bytes | Expected | Returns `"(no output)"` — correct for a killed-early command. |
| Shell-side behavior changes | None | Follow-up is Node-side only. |
| Regresses Linux streaming | None | Helper runs only on fallback paths, identical across platforms. |

## Post-deploy validation: NOCTURNE 003 → 004

After landing both commits on this branch, the first NOCTURNE re-run (003) appeared to show that **leaks were unchanged** despite the fix. Investigation revealed the cause: Evalgate invokes APEX through the published bin (`bin/pensar.js → build/cli.js`), which is a static bundle produced by `bun run build`. That bundle had not been regenerated since March, so the fix was present in source but not in the running binary. There is no `postinstall` / pre-test rebuild hook, so any local change to `src/` is invisible to Evalgate until `bun run build` runs.

After `bun run build` (`build/cli-242p2y3r.js` now contains 6 references to `extractFallbackStdout` — 1 declaration + 5 fallback call sites — and all bundle chunks are timestamped post-fix), NOCTURNE 004 was run on the same target with the same prompt.

### NOCTURNE 003 vs 004

| Metric | NOCTURNE 003 (stale bundle, fix not deployed) | NOCTURNE 004 (rebuilt bundle, fix deployed) |
|---|---|---|
| `execute_command` observations | 63 | 71 |
| Steps containing `execute_command` | (similar) | 37 |
| Steps issuing 2+ parallel `execute_command` | (similar) | **31 / 37 (84%)** |
| Steps issuing 3 parallel | (similar) | 3 |
| Marker leaks in tool stdout | 2 | **1** |
| Leak shape | own-nonce + cross-nonce mixed | **cross-nonce only** |

The single residual leak in 004 is **not** the same bug as commit 2 of this PR. It surfaces a separate, lower-frequency bug in `PersistentShell` that the helper is structurally unable to fix.

## Known limitation: concurrent-call output bleed (separate bug, follow-up PR)

The one residual leak in NOCTURNE 004 is forensic proof of a distinct race condition in `PersistentShell.execute()` — concurrent commands sharing a single shell can cross-contaminate each other's stdout *and stderr* buffers.

### Evidence

Step 4 of `attack-surface-agent` issues two parallel `execute_command` calls:

| | Call A `toolu_013gEBbqhmGpnRSqMJo9kYCt` | Call B `toolu_011teCwm4UHUs9t4SV1G8E9o` |
|---|---|---|
| description | "Find available wordlists" | "Test Flask-specific paths and more endpoint patterns" |
| command | `find / -name "*.txt" -path "*/wordlist*" …; find …; find …; find …` | `for path in /meridian /auth … /playground; do curl … done` |
| timeout | **10 s** | **120 s** |
| `result.success` | false | false |
| `result.error` | `Command timed out` | `Command timed out` |
| `result.stdout` | `(no output)` ← clean ✓ | `__APEX_8b35a2965745ce81___CUTOVER\n__APEX_8b35a2965745ce81___EXIT_137\n` ← LEAK |
| `result.stderr` | `""` | `bash: line 126: 4264 Terminated: 15 find / -name "*.txt" -path "*/wordlist*" …\n4416 Killed: 9   find / -name "*.txt" -path "*/dirb*" …\n4421 Terminated: 15 find / -name "*.txt" -path "*/dirbuster*" …\n5878 Killed: 9   find / -name "*.txt" -path "*/SecLists*" …` |

Call B never spawned a `find`. Yet Call B's stderr contains the kernel-termination messages for **Call A's** four `find` processes, and Call B's stdout contains a full wrapper-output skeleton (`CUTOVER\nEXIT_137\n` with empty body) that exactly matches Call A's behavior — `find`'s stdout was empty because `2>/dev/null` redirected its only output (stderr) away, `__APEX_EC=$?` captured 137 from the `SIGKILL` escalation, and the cleanup phase emitted an empty `cat` between the markers.

### Root cause

`PersistentShell.execute()` has no per-shell command queue:

```ts
// src/core/agents/offSecAgent/tools/persistentShell.ts:408
this.current = pending;
this.pendingCancel = pending.resolve;
```

When the agent issues parallel tool calls in a single Claude turn, both `execute()` invocations land in the same JS tick. They run synchronously to the `return new Promise(…)`:

1. Each call writes its wrapper to `proc.stdin` immediately.
2. Each call assigns `this.current = pending` — last writer wins.

Bash drains stdin serially, executing wrapper A then wrapper B. **All bash output flows to whichever pending is currently `this.current`** — i.e. pending B for both wrappers. When wrapper A's cleanup phase emits its CUTOVER + cat + EXIT (under nonce A), those bytes accumulate in pending **B**'s `streamedStdout`. Pending A's `streamedStdout` stays empty for its entire lifetime, so its fallback returns `(no output)` (clean). Pending B's fallback finds A's markers in its buffer, can't match its own (B's) nonce, and — correctly per its contract — returns the buffer verbatim. The helper does not strip foreign nonces, by design: those bytes could be legitimate user output that happens to contain a marker-shaped string.

The same bleed silently corrupts **stderr**: pending B's stderr contains pending A's `find` kill messages, with no marker pattern to flag the contamination. The model cannot tell these messages aren't about its own curl loop. This is a correctness hazard for agent reasoning that the marker stripping cannot detect.

### Why so rare in 004 (only 1 leak in 31 parallel-call steps)

Most parallel batches resolve cleanly because:

- The shorter-timeout call's timer fires first, forces it to fall back to `(no output)` while bash is still working through the queue.
- Bash continues executing queued wrappers; the **last** pending (the one `this.current` points to) eventually sees its own CUTOVER/EXIT and the happy-path parser strips them clean.

The leak surfaces only when the first wrapper finishes (or is killed) after a different pending has been registered as `this.current`, **and** the holder of `this.current` also fails to complete its own happy path within the fallback window. Narrow window — but with 31 parallel-call steps in one trace, it hit at least once.

### Why this is a separate PR

The fix is **serialize `execute()` per shell with a FIFO promise chain** — small, mechanical change that makes each pending see only its own wrapper's bytes. The marker-stripping helper's correctness assumption (one command's nonces are in one command's buffer) becomes structurally guaranteed.

That is a behavioral change (parallel `execute_command` calls become serial through a shared shell) and deserves its own review and its own NOCTURNE pass. Holding it out of this PR keeps the wins isolated:

- This PR: silent stdout-loss eliminated; own-nonce marker leaks eliminated; one structural class of bug remains, fully diagnosed.
- Follow-up PR: cross-command output bleed eliminated.

Both ship before the next eval cycle. The follow-up is tracked in #645 so reviewers of this PR don't have to evaluate two semantically distinct fixes at once.

## Rollback

Touches exactly one file plus its test. No data-format change, no migration. Single \`git revert\`; zero blast radius.

## Test plan

- [x] CI passes (Lint / TypeCheck / Test / Build)
- [ ] Manual TUI smoke on a Mac: operator mode against a local target, confirm recon returns real output instead of \`\"(no output)\"\`
- [ ] Manual CLI smoke on a Mac: \`bun run src/cli.ts pentest --target ...\` inside a trivial challenge
- [x] Re-run NOCTURNE via Evalgate on a Mac (run 004) — silent-truncation incidents eliminated; see *Post-deploy validation* table above
- [x] Confirm marker leaks have collapsed to the single residual concurrent-call bleed case — diagnosed and tracked for a separate follow-up PR
- [ ] Optional: manual run inside Alpine container to confirm correctness on musl

Made with [Cursor](https://cursor.com)

---

> [!NOTE]
> **Medium Risk**
> Touches core command-execution plumbing used by the offsec agent; bugs could cause missing/duplicated output or incorrect exit handling across platforms, though changes are well-covered by targeted regression tests.
> 
> **Overview**
> Makes `PersistentShell` return **platform-independent, authoritative stdout** by always running a post-`tail` cutover marker plus `cat` phase, and splitting output into pre-cutover *best-effort streaming* vs post-cutover *authoritative capture*.
> 
> Adds `extractFallbackStdout` and routes timeout/abort/cancel/close fallback resolves through it to prevent `__APEX_*` cutover/exit markers leaking into tool stdout when fallback timers win event-loop races.
> 
> Expands `persistentShell.test.ts` with regressions for macOS/blocked-buffer `tail` (trivial commands, 8KiB boundary) and marker-leak guards, plus direct unit tests for `extractFallbackStdout`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 8f14c3a524ccde8a04c8926b386a28cbfb72b2e2. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/core/agents/offSecAgent/tools/persistentShell.test.ts [hunk 1]
```diff
@@ -235,12 +289,52 @@ describe("PersistentShell — long-running stability", () => {
...
+   * Useful as a second line of defense against marker leaks even though
+   * the real-world trigger (busy event loop, setTimeout fires before
+   * onStdoutData processes the exit-marker chunk) is covered by the
+   * dedicated `extractFallbackStdout` unit tests below.
+   */
+  it("does not leak nonce markers on timeout", async () => {
+    const shell = make();
+    const r = await shell.execute(`printf 'hello\\n'; sleep 10`, 1);
+    expect(r.exitCode).toBe(124);
+    expect(r.stdout).not.toMatch(/__APEX_[0-9a-f]+__(CUTOVER|EXIT_)/);
```
[the remark this answers is not in the document]
Comment (kylejryan) [reply src/core/agents/offSecAgent/tools/persistentShell.test.ts]: @cursoragent Obliterate this bug

File: src/core/agents/offSecAgent/tools/persistentShell.ts [hunk 2]
```diff
@@ -74,6 +95,61 @@ function hasStdbuf(): boolean {
...
+ * trigger requires a busy event loop which is hard to force in a
+ * deterministic integration test.
+ */
+export function extractFallbackStdout(cmd: {
+  authoritativeStdout: string;
+  streamedStdout: string;
+  cutoverMarker: string;
+  exitMarkerPrefix: string;
+}): string {
+  if (cmd.authoritativeStdout) return cmd.authoritativeStdout;
```
[the remark this answers is not in the document]
Comment (kylejryan) [reply src/core/agents/offSecAgent/tools/persistentShell.ts]: @cursoragent Fix pls

File: src/core/agents/offSecAgent/tools/persistentShell.ts [hunk 3]
```diff
@@ -74,6 +95,69 @@ function hasStdbuf(): boolean {
...
+  cutoverMarker: string;
+  exitMarkerPrefix: string;
+}): string {
+  if (cmd.authoritativeStdout) {
+    let s = cmd.authoritativeStdout;
+    const exitIdx = s.indexOf(cmd.exitMarkerPrefix);
+    if (exitIdx !== -1) {
+      s = s.substring(0, exitIdx);
+      if (s.endsWith("\n")) s = s.substring(0, s.length - 1);
+    }
```
[the remark this answers is not in the document]
Comment (kylejryan) [reply src/core/agents/offSecAgent/tools/persistentShell.ts:139]: @cursoragent Fix

**Review by jorgeraad — APPROVED** [review]
